Description: Deprecate pam_unix' explicit "usergroups" option and instead read it from /etc/login.def's "USERGROUP_ENAB" option if umask is only defined there. This restores compatibility with the pre-PAM behaviour of login. See https://blueprints.launchpad.net/ubuntu/+spec/umask-to-0002.
Author: Martin Pitt <martin.pitt@ubuntu.com>
Bug-Debian: http://bugs.debian.org/583958

=== modified file 'modules/pam_umask/pam_umask.c'
Index: pam-1.1.3/modules/pam_umask/pam_umask.c
===================================================================
--- pam-1.1.3.orig/modules/pam_umask/pam_umask.c	2016-03-17 13:11:16.763228404 -0500
+++ pam-1.1.3/modules/pam_umask/pam_umask.c	2016-03-17 13:11:16.759228425 -0500
@@ -87,7 +87,7 @@ parse_option (const pam_handle_t *pamh,
 }
 
 static char *
-search_key (const char *filename)
+search_key (const char *filename, const char *key)
 {
   FILE *fp;
   char *buf = NULL;
@@ -142,7 +142,7 @@ search_key (const char *filename)
         while (isspace ((int)*cp) || *cp == '=')
           ++cp;
 
-      if (strcasecmp (tmp, "UMASK") == 0)
+      if (strcasecmp (tmp, key) == 0)
 	{
 	  retval = strdup (cp);
 	  break;
@@ -159,15 +159,34 @@ static int
 get_options (const pam_handle_t *pamh, options_t *options,
 	     int argc, const char **argv)
 {
+  char *result;
+
   memset (options, 0, sizeof (options_t));
   /* Parse parameters for module */
   for ( ; argc-- > 0; argv++)
     parse_option (pamh, *argv, options);
 
   if (options->umask == NULL)
-    options->umask = search_key (LOGIN_DEFS);
+    {
+      options->umask = search_key (LOGIN_DEFS, "UMASK");
+      /* login.defs' USERGROUPS_ENAB will modify the UMASK setting there by way
+       * of usergroups; but we don't want it to influence umask definitions
+       * from other places (like GECOS). This restores compatibility with
+       * shadow from the pre-PAM age.
+       */ 
+      if (options->umask != NULL)
+	{
+	  result = search_key (LOGIN_DEFS, "USERGROUPS_ENAB");
+	  if (result != NULL)
+	    {
+	      options->usergroups = (strcasecmp (result, "yes") == 0);
+	      free (result);
+	    }
+	}
+    }
+
   if (options->umask == NULL)
-    options->umask = search_key (LOGIN_CONF);
+    options->umask = search_key (LOGIN_CONF, "UMASK");
 
   return 0;
 }
Index: pam-1.1.3/modules/pam_umask/pam_umask.8.xml
===================================================================
--- pam-1.1.3.orig/modules/pam_umask/pam_umask.8.xml	2016-03-17 13:11:16.763228404 -0500
+++ pam-1.1.3/modules/pam_umask/pam_umask.8.xml	2016-03-17 13:11:16.759228425 -0500
@@ -73,7 +73,8 @@
         </listitem>
         <listitem>
           <para>
-            UMASK entry from /etc/login.defs
+            UMASK entry from /etc/login.defs (influenced by USERGROUPS_ENAB in
+	    /etc/login.defs)
           </para>
         </listitem>
       </itemizedlist>
@@ -118,6 +119,11 @@
               If the user is not root and the username is the same as
               primary group name, the umask group bits are set to be the
               same as owner bits (examples: 022 -> 002, 077 -> 007).
+	      Note that using this option explicitly is discouraged. pam_umask
+	      enables this functionality by default if /etc/login.defs enables
+	      USERGROUPS_ENAB, and the umask is not set explicitly in other
+	      places than /etc/login.defs (this is compatible with login's
+	      behaviour without PAM).
             </para>
           </listitem>
         </varlistentry>
Index: pam-1.1.3/modules/pam_umask/pam_umask.8
===================================================================
--- pam-1.1.3.orig/modules/pam_umask/pam_umask.8	2016-03-17 13:11:16.763228404 -0500
+++ pam-1.1.3/modules/pam_umask/pam_umask.8	2016-03-17 13:11:16.759228425 -0500
@@ -1,13 +1,13 @@
 '\" t
 .\"     Title: pam_umask
 .\"    Author: [see the "AUTHOR" section]
-.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
-.\"      Date: 09/19/2013
+.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
+.\"      Date: 03/17/2016
 .\"    Manual: Linux-PAM Manual
 .\"    Source: Linux-PAM Manual
 .\"  Language: English
 .\"
-.TH "PAM_UMASK" "8" "09/19/2013" "Linux-PAM Manual" "Linux\-PAM Manual"
+.TH "PAM_UMASK" "8" "03/17/2016" "Linux-PAM Manual" "Linux\-PAM Manual"
 .\" -----------------------------------------------------------------
 .\" * Define some portability stuff
 .\" -----------------------------------------------------------------
@@ -101,7 +101,7 @@ UMASK= entry from /etc/default/login
 .sp -1
 .IP \(bu 2.3
 .\}
-UMASK entry from /etc/login\&.defs
+UMASK entry from /etc/login\&.defs (influenced by USERGROUPS_ENAB in /etc/login\&.defs)
 .RE
 .sp
 .SH "OPTIONS"
@@ -119,7 +119,7 @@ Don\*(Aqt print informative messages\&.
 .PP
 \fBusergroups\fR
 .RS 4
-If the user is not root and the username is the same as primary group name, the umask group bits are set to be the same as owner bits (examples: 022 \-> 002, 077 \-> 007)\&.
+If the user is not root and the username is the same as primary group name, the umask group bits are set to be the same as owner bits (examples: 022 \-> 002, 077 \-> 007)\&. Note that using this option explicitly is discouraged\&. pam_umask enables this functionality by default if /etc/login\&.defs enables USERGROUPS_ENAB, and the umask is not set explicitly in other places than /etc/login\&.defs (this is compatible with login\*(Aqs behaviour without PAM)\&.
 .RE
 .PP
 \fBumask=\fR\fB\fImask\fR\fR
